From 5eab1f32f0bb922ea61f0a253621112f5f0c881e Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Tue, 30 Dec 2025 22:22:22 +0800
Subject: [PATCH 51/63] loonggpu: backlight: get PWM correctly and skip GPIO
 for now

This should make backlight control at least usable with AOSC 6.16.1-rc1
kernel.

Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 loonggpu/loonggpu_backlight.c | 38 ++++++++++++++++++++++++++++++++---
 1 file changed, 35 insertions(+), 3 deletions(-)

diff --git a/loonggpu/loonggpu_backlight.c b/loonggpu/loonggpu_backlight.c
index e24421a..f1bbdea 100644
--- a/loonggpu/loonggpu_backlight.c
+++ b/loonggpu/loonggpu_backlight.c
@@ -11,8 +11,12 @@
 
 static bool loonggpu_backlight_get_hw_status(struct loonggpu_backlight *ls_bl)
 {
+#ifdef LG_GPIO_WORKS
 	return (gpio_get_value(ls_bl->pin_vdd) && gpio_get_value(ls_bl->pin_en)
 		&& pwm_is_enabled(ls_bl->pwm));
+#else
+	return pwm_is_enabled(ls_bl->pwm);
+#endif
 }
 
 static int loonggpu_lcd_ctrl_open_detect(struct loonggpu_backlight *ls_bl)
@@ -23,18 +27,22 @@ static int loonggpu_lcd_ctrl_open_detect(struct loonggpu_backlight *ls_bl)
 
 static int loonggpu_lcd_ctrl_vdd(struct loonggpu_backlight *ls_bl, unsigned int delay_ms, bool open)
 {
+#ifdef LG_GPIO_WORKS
 	gpio_direction_output(ls_bl->pin_vdd, 1);
 	gpio_set_value(ls_bl->pin_vdd, open);
 	mdelay(delay_ms);
+#endif
 
 	return 0;
 }
 
 static int loonggpu_lcd_ctrl_en(struct loonggpu_backlight *ls_bl, unsigned int delay_ms, bool open)
 {
+#ifdef LG_GPIO_WORKS
 	gpio_direction_output(ls_bl->pin_en, 1);
 	gpio_set_value(ls_bl->pin_en, open);
 	mdelay(delay_ms);
+#endif
 
 	return 0;
 }
@@ -258,10 +266,23 @@ static int loonggpu_backlight_hw_request_init(struct loonggpu_backlight *ls_bl)
 	bool pwm_enable_default;
 	struct pwm_state state;
 	struct loonggpu_device *adev = ls_bl->driver_private;
-	char pwm_name[16];
 
-	sprintf(pwm_name, "pwm%d", ls_bl->pwm_id);
-	ls_bl->pwm = lg_pwm_request(adev->ddev->dev, pwm_name, ls_bl->pwm_id, "Loongson_bl");
+	if (ls_bl->pwm_id != 3) {
+		DRM_ERROR("Must use PWM3 for backlight\n");
+		goto ERROR_PWM;
+	}
+
+	/*
+	 * Note that we must patch the kernel ACPI initialization code to
+	 * register the name loonggpu_backlight using pwm_add_table to make
+	 * this work.  We cannot call pwm_add_table in this module as
+	 * pwm_add_table is only intended for board initialization and not
+	 * exported.  Even some dirty hack might allow us to call it from a
+	 * module, doing so would still cause a deadlock.
+	 *
+	 * See https://github.com/AOSC-Tracking/linux/commit/6ce646344dfb.
+	 */
+	ls_bl->pwm = pwm_get(adev->ddev->dev, "gsgpu_backlight");
 
 	if (IS_ERR(ls_bl->pwm)) {
 		DRM_ERROR("Failed to get the pwm chip\n");
@@ -283,6 +304,12 @@ static int loonggpu_backlight_hw_request_init(struct loonggpu_backlight *ls_bl)
 	if (pwm_enable_default)
 		pwm_enable(ls_bl->pwm);
 
+	/*
+	 * FIXME: The LCD backlight enable/disable and LCD power
+	 * enable/disable support is removed for now.  The upstream
+	 * kernel seems not able to find GPIO_LCD_VDD and GPIO_LCD_EN.
+	 */
+#ifdef LG_GPIO_WORKS
 	ret = gpio_request(GPIO_LCD_VDD, "GPIO_VDD");
 	if (ret) {
 		DRM_ERROR("EN request error!\n");
@@ -298,13 +325,16 @@ static int loonggpu_backlight_hw_request_init(struct loonggpu_backlight *ls_bl)
 	/* gpio init */
 	gpio_direction_output(GPIO_LCD_VDD, 1);
 	gpio_direction_output(GPIO_LCD_EN, 1);
+#endif
 
 	return ret;
 
+#ifdef LG_GPIO_WORKS
 ERROR_EN:
 	gpio_free(GPIO_LCD_VDD);
 ERROR_VDD:
 	lg_pwm_free(ls_bl->pwm);
+#endif
 ERROR_PWM:
 	return -ENODEV;
 }
@@ -381,7 +411,9 @@ static struct loonggpu_lcd_ctrl * loonggpu_lcd_ctrl_init(struct loonggpu_device
 		loonggpu_lcd_ctrl_init_sequence_list(adev, lcd_ctrl, lcd_ctrl_res, LCD_CTRL_ACTION_CLOSE);
 		loonggpu_lcd_ctrl_init_sequence_list(adev, lcd_ctrl, lcd_ctrl_res, LCD_CTRL_ACTION_OPEN);
 
+#ifdef LG_GPIO_WORKS
 		lcd_ctrl->gpio_detect_open_pin = lcd_ctrl_res->gpio_detect_open_pin;
+#endif
 		lcd_ctrl->gpio_detect_open_timer = lcd_ctrl_res->gpio_detect_open_timer;
 		lcd_ctrl->gpio_detect_open_polarity = lcd_ctrl_res->gpio_detect_open_polarity;
 		lcd_ctrl->gpio_ctrl_vdd = lcd_ctrl_res->gpio_ctrl_vdd;
-- 
2.52.0

